### ISSUE TITLE
Targets don't work with custom properties on HTMLElement

### ISSUE THREAD

**Issue by mlwyatt** [ajaishankar/stimulus-typescript#1]
I'm impressed you managed to get this working, but I knew it was going to take someone who knew TS better than me. I've only been doing it less than a year.

I'm definitely going to be adding this package to my app. However, some of my `targets` are elements with custom properties added in `connect` and my attempt to add those custom properties with `ObjectAs` isn't quite working.

<details>
<summary>Controller A</summary>

```ts
import SlimSelect from 'slim-select';

class SlimWithToggle extends SlimSelect {
  toggle(disabled: boolean): void {
    if ( this.settings.disabled !== disabled ) {
      return;
    }

    if ( disabled ) {
      this.disable();
    }
    else {
      this.enable();
    }
  }
}

interface HTMLElementWithSlimSelect extends HTMLSelectElement {
  slim?: SlimWithToggle;
}

export default class extends Controller<HTMLElementWithSlimSelect> {
  slimSelect: SlimWithToggle | undefined;

  connect(): void {
    ...
    this.slimSelect = new SlimWithToggle(options);
  }
}
```
</details>

<details>
<summary>Controller B</summary>

```ts
import type { HTMLElementWithSlimSelect } from './controller_a';
import { Controller } from '@hotwired/stimulus';
import { ObjectAs, Typed } from 'stimulus-typescript';

const targets = {
  selectField: ObjectAs<HTMLElementWithSlimSelect>,
  theTextField: HTMLSpanElement,
  idField: HTMLInputElement,
};

class MyNewController extends Typed(Controller<HTMLElement>, { targets }) {
  connect(): void {
    // Just to test compiler
    console.log(this.idFieldTarget);
    console.log(this.idFieldTargets);
    console.log(this.hasIdFieldTarget);
    console.log(this.theTextFieldTarget);
    console.log(this.theTextFieldTargets);
    console.log(this.hasTheTextFieldTarget);
    console.log(this.selectFieldTarget);
    console.log(this.selectFieldTarget.slim?.toggle);
    console.log(this.selectFieldTargets);
    console.log(this.hasSelectFieldTarget);
  }
}
```
</details>

Removing all things `selectField` from controller B compiles perfectly. However, when I add them back, the compiler complains about the other targets needing to be accessed via index signature and `HTMLElementWithSlimSelect` doesn't extend `Record<string, unknown>`

<details>
<summary>TS Output</summary>

```sh
controller_b.ts(6,25): error TS2344: Type 'HTMLElementWithSlimSelect' does not satisfy the constraint 'Record<string, unknown>'.
  Index signature for type 'string' is missing in type 'HTMLElementWithSlimSelect'.
controller_b.ts(11,64): error TS2322: Type '{ selectField: typeof ObjectAs; theTextField: { new (): HTMLSpanElement; prototype: HTMLSpanElement; }; idField: { new (): HTMLInputElement; prototype: HTMLInputElement; }; }' is not assignable to type 'TargetDefinitionMap'.
  Property 'selectField' is incompatible with index signature.
    Type 'typeof ObjectAs' is not assignable to type '{ new (): Element; prototype: Element; }'.
      Types of property 'prototype' are incompatible.
        Type 'ObjectAs<any>' is missing the following properties from type 'Element': attributes, classList, className, clientHeight, and 163 more.
controller_b.ts(13,22): error TS4111: Property 'idFieldTarget' comes from an index signature, so it must be accessed with ['idFieldTarget'].
controller_b.ts(14,22): error TS4111: Property 'idFieldTargets' comes from an index signature, so it must be accessed with ['idFieldTargets'].
controller_b.ts(15,22): error TS4111: Property 'hasIdFieldTarget' comes from an index signature, so it must be accessed with ['hasIdFieldTarget'].
controller_b.ts(16,22): error TS4111: Property 'theTextFieldTarget' comes from an index signature, so it must be accessed with ['theTextFieldTarget'].
controller_b.ts(17,22): error TS4111: Property 'theTextFieldTargets' comes from an index signature, so it must be accessed with ['theTextFieldTargets'].
controller_b.ts(18,22): error TS4111: Property 'hasTheTextFieldTarget' comes from an index signature, so it must be accessed with ['hasTheTextFieldTarget'].
controller_b.ts(19,22): error TS4111: Property 'selectFieldTarget' comes from an index signature, so it must be accessed with ['selectFieldTarget'].
controller_b.ts(20,17): error TS2532: Object is possibly 'undefined'.
controller_b.ts(20,22): error TS4111: Property 'selectFieldTarget' comes from an index signature, so it must be accessed with ['selectFieldTarget'].
controller_b.ts(20,40): error TS2339: Property 'slim' does not exist on type 'Element'.
controller_b.ts(21,22): error TS4111: Property 'selectFieldTargets' comes from an index signature, so it must be accessed with ['selectFieldTargets'].
controller_b.ts(22,22): error TS4111: Property 'hasSelectFieldTarget' comes from an index signature, so it must be accessed with ['hasSelectFieldTarget'].
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
</details>

**Comment by ajaishankar** [comment]
Thanks @mlwyatt!

For someone with less than a year of Typescript you're not doing bad either 😄 

`ObjectAs` is only meant to type a controller data `value` (ObjectAs instead of Object) and won't work on targets.

Thanks for sharing - will add a custom `Target` that should fix this.

```ts
const targets = {
  selectField: Target<HTMLElementWithSlimSelect>,
  theTextField: HTMLSpanElement,
  idField: HTMLInputElement,
};

```

**Comment by ajaishankar** [comment]
Have released v0.1.2

Aliased `ObjectAs` to `Object_`
Renamed `TargetAs` to `Target` and also removed restriction on Target needing to be an element.
```ts
interface CustomSelect { search: string }

const targets = {
  select: Target<CustomSelect>
}
```
